torch/csrc/distributed/rpc/init.cpp

```diff
@@ -333,23 +346,71 @@ PyObject* rpc_init(PyObject* /* unused */) {
   // pythonRpcHandler is cleaned up in shutdown(), after
   // shutdown(), python objects returned from rpc python call can not be
   // resolved.
-  auto future = shared_ptr_class_<FutureIValue>(module, "Future")
-                    .def(
-                        "wait",
-                        [&](FutureIValue& fut) {
-                          auto& pythonRpcHandler =
-                              PythonRpcHandler::getInstance();
-                          const auto& value = fut.wait();
-                          pybind11::gil_scoped_acquire ag;
-                          auto obj = torch::jit::toPyObject(value);
-                          pythonRpcHandler.handleException(obj);
-                          return obj;
-                        },
-                        py::call_guard<py::gil_scoped_release>(),
-                        R"(
+  shared_ptr_class_<FutureIValue>(module, "Future")
+      .def(
+          "wait",
+          [&](FutureIValue& fut) {
+            auto& pythonRpcHandler = PythonRpcHandler::getInstance();
+            const auto& value = fut.wait();
+            pybind11::gil_scoped_acquire ag;
+            auto obj = torch::jit::toPyObject(value);
+            pythonRpcHandler.handleException(obj);
+            return obj;
+          },
+          py::call_guard<py::gil_scoped_release>(),
+          R"(
 Wait on future to complete and return the object it completed with.
 If the future completes with an error, an exception is thrown.
-              )");
+          )")
+      .def(
+          "_then",
+          [&](FutureIValue& fut, py::function cb) {
+            // We need this an additional layer of wrapper here to guard the
+            // destruction of the py::function object. Because, the
+            // FutureIValue owns a reference to the py::function in its
+            // callback vector, but FutureIValue does not acquire GIL on
+            // destruction and it does not need to do so either. FutureIValue
+            // only acquires GIL if the IValue holds a py::object. However,
+            // nothing prevents applications to insert py::function as callbacks
+            // non-py::object FutureIValue. For example, the application can
+            // call a TorchScript function using rpc_async, and then append
+            // py::function as a callback. Hence, we use PythonFunction wrapper
+            // here to explicitly acquire GIL and decouple value destruction and
+            // callback destruction.
+            PythonFunction pf(std::move(cb));
+            return fut.then<IValue>([pf](const FutureIValue& fut) -> IValue {
```

Comment:
nit: I think the `const FutureIValue& fut` shadows the self argument `FutureIValue& fut` on 367?

Reply:
ahh, right, let me change it to a different name.